### DISCUSSION TITLE
Kill spawned processes on App close from Javascript Code

### DISCUSSION THREAD

**Discussion by touhidurrr** [tauri-apps/tauri#5870] [Q&A] [answered]
Hi! I am using Tauri for a while to write production apps. I am not being able to close spawned processes when the app is closed. How do I do it?

Note I tried listening to [`WINDOW_DESTROYED`](https://tauri.app/v1/api/js/event#event.TauriEvent.WINDOW_DESTROYED) event from JavaScript API, using the browsers [`onbeforeunload`](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event) event, etc but none worked.

Also, I tried using this code to listen to all events but it seems like the code cannot detect some events like [`WINDOW_CLOSE_REQUESTED`](https://tauri.app/v1/api/js/event#event.TauriEvent.WINDOW_CLOSE_REQUESTED), [`WINDOW_CREATED`](https://tauri.app/v1/api/js/event#event.TauriEvent.WINDOW_CREATED), [`WINDOW_DESTROYED`](https://tauri.app/v1/api/js/event#event.TauriEvent.WINDOW_DESTROYED), etc from JavaScript API.

```ts
import { writeTextFile } from "@tauri-apps/api/fs";
import { listen, TauriEvent } from "@tauri-apps/api/event";

Object.entries(TauriEvent).forEach(([key, value]) => {
  listen<string>(value, async (event) => {
    console.log("Closing App..");
    writeTextFile(key, `${key} on ${new Date().toLocaleString()}\n${JSON.stringify(event, null, 2)}`);
  });
});
```

Only these 4 events were logged after much testing:
![image](https://user-images.githubusercontent.com/46617994/208453395-f9615e7e-9fc2-41b4-be94-0da9d250b432.png)

Can anyone tell me the correct way to do it?

**EDIT:** [`appWindow.onCloseRequested`](https://tauri.app/v1/api/js/window#oncloserequested) worked after dynamically importing `@tauri-apps/api/window`.

**Comment by FabianLars** [comment, marked as answer]
hmm, maybe try the window specific `listen` methods (like `appWindow.listen`), which also has a helper function: https://tauri.app/v1/api/js/window#oncloserequested

> Hi! I am using Tauri for a while to write production apps. I am not being able to close spawned processes when the app is closed. How do I do it?

It depends. If it's spawned as a sidecar then Tauri _should_ handle it itself. The problem is that some (sidecar) apps spawn multiple processes without cleaning up after themselves. This is currently only solvable in Rust, see https://github.com/tauri-apps/tauri/discussions/3273 for example.
If we're talking about "normal" Commands (not sidecars) then you need to handle it yourself.

**Comment by touhidurrr** [comment]
@FabianLars I don't know Rust. I am a FullStack JS/TS developer. I am using the `@tauri-apps/api/shell` api to spawn command and kill them while Tauri throws some kind of app close event. For that I tried using `@tauri-apps/api/event` and as you can see that is not working.

What do you recommend me to do in this situation?

**Comment by touhidurrr** [comment]
> hmm, maybe try the window specific listen methods (like appWindow.listen), which also has a helper function: https://tauri.app/v1/api/js/window#oncloserequested

tried that also in the dev mode but got the following error:
![image](https://user-images.githubusercontent.com/46617994/208475458-bdbb0206-6125-49ed-834b-c2639651c3f6.png)

**Reply by FabianLars** [reply]
Sounds like you're using a SSR framework? In which case you need to change the appWindow import to a dynamic import (the `await import('')` thingies)

**Reply by touhidurrr** [reply]
ok, dynamic imports worked. but before you, my friend came up with the solution by trial and error. thanks for helping though anyways.
